### PR TITLE
endpoints for follow and unfollow

### PR DESCRIPTION
**Pull request by Lakshya021**
added the endpoints for follow and unfollow
haven't been able to test as of yet.

Update : Tested & Working
Screen Shot 2022-07-13 at 4 52 35 PM
Screen Shot 2022-07-13 at 4 53 47 PM

Real time updates happening in the MongoDB database.

Also added the getUser Endpoint

### CODE REVIEW

File: server/controllers/userControllers.ts [hunk 1]
```diff
@@ -1,24 +1,40 @@
...
+            if(!userToBeFollowed.followers.includes(req.body.id)){
+                console.log("comes inside");
+                await userToBeFollowed.updateOne({ $push: { followers : req.body.id }});
+                console.log("followers list edited");
+                await userFollowing.updateOne({$push: { followings : req.params.id }})
+                console.log("following list updated");
                 res.status(200).json("user has been followed");
             }
             else{
                 res.status(403).json("you allready follow this user");
```
Comment (danielcrookall) [inline server/controllers/userControllers.ts:33]: typo in "already"

File: server/router.ts [hunk 2]
```diff
@@ -9,11 +11,15 @@ const router = express.Router();
  * AUTHENTICATION ENDPOINTS
  */
 
+router.get("/api/users/:id", getUser)
 router.post("/api/users/register", registerUser);
 router.post("/api/users/login", loginUser);
 router.post("/api/users/forgot-password", forgotPassword);
 router.put("/api/users/reset-password/:resetToken", resetPassword);
 
+router.put("/api/users/:id/follow", followUser);
```
Comment (swongg) [inline server/router.ts:20]: Just an FYI -- It's considered bad practice to use verbs (i.e. follow, register) for our RESTFUL apis and instead, we should be sticking with nouns. I noticed that we have this in a lot of our routes which is fine for the sake of this project :)

Source: https://docs.microsoft.com/en-us/azure/architecture/best-practices/api-design
 _When possible, resource URIs should be based on nouns (the resource) and not verbs (the operations on the resource)._

File: server/controllers/userControllers.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    if(req.params.id !== req.body.id){
+        // params is storing the id of the user who is to Followed
+        // body is storing the id of the user who is doing the action of following
+        try{
+            console.log("following a user");
+            const userToBeFollowed = await User.findById(req.params.id);
+            const userFollowing = await User.findById(req.body.id);
+            console.log("storing users");
+            if(!userToBeFollowed.followers.includes(req.body.id)){
+                console.log("comes inside");
```
Comment (swongg) [inline server/controllers/userControllers.ts:25]: probably should remove the log statements unless you really need them. especially the "comes inside" one

File: server/controllers/userControllers.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+      const { password, updatedAt, ...other } = user._doc;
+      res.status(200).json(other);
+    } catch (err) {
+      res.status(500).json(err);
+    }
+}
+  
+
+export const followUser = async (req: Request, res: Response) => {
+    if(req.params.id !== req.body.id){
```
Comment (swongg) [inline server/controllers/userControllers.ts:16]: 👍

File: server/router.ts [hunk 5]
```diff
@@ -9,11 +11,15 @@ const router = express.Router();
...
+router.put("/api/users/:id/unfollow", unfollowUser);
```
Comment (swongg) [inline server/router.ts:21]: Do we need both ```unfollow``` and ```follow``` endpoints or can they be combined into one?  Perhaps the ```PUT``` request can contain some type of indicator to specify of it's a _follow_ or an _unfollow_ ?

**Review by swongg — CHANGES_REQUESTED** [review]

File: server/controllers/userControllers.ts [hunk 6]
```diff
@@ -0,0 +1,64 @@
+import User from "../models/user.model";
+import { Request, Response } from 'express';
+
+export const getUser = async (req: Request, res: Response) => {
+    try {
+      const user = await User.findById(req.params.id);
+      const { password, updatedAt, ...other } = user._doc;
+      res.status(200).json(other);
```
[the remark this answers is not in the document]
Comment (Lakshya021) [reply server/controllers/userControllers.ts:8]: Yeah, exactly what you said!

File: server/router.ts (hunk 2 above)
Comment (Lakshya021) [reply server/router.ts:20]: Good point Steven! I have modified it in the new commit.

File: server/router.ts [hunk 7]
```diff
@@ -18,8 +18,8 @@ router.post("/api/users/login", loginUser);
 router.post("/api/users/forgot-password", forgotPassword);
 router.put("/api/users/reset-password/:resetToken", resetPassword);
 
-router.put("/api/users/:id/follow", followUser);
-router.put("/api/users/:id/unfollow", unfollowUser);
+router.put("/api/users/:id", editUser);
+// router.put("/api/users/:id", unfollowUser);
```
Comment (swongg) [inline server/router.ts:22]: please remove this commented out cde

File: server/controllers/userControllers.ts [hunk 8]
```diff
@@ -128,55 +128,96 @@ export const getUser = async (req: Request, res: Response) => {
...
+            error: {
+                userToBeUnfollowedId: req.params.id
+            }
+        });
       }
 }
 
 
+
+export const editUser = (req:Request, res: Response) => {
```
Comment (swongg) [inline server/controllers/userControllers.ts:203]: I was thinking for ```editUser```, we shouldn't have to specify the action type.
If we did it this way, we would essentially have to keep adding new cases to our switch statement everytime we want to modify a different field.

**Pull request merged**